brainiak/fcma/cython_blas.pyx

```diff
@@ -0,0 +1,200 @@
+    A = py_a
+    cdef float[:, :, ::1] C
+    C = py_c
+    blas.sgemm(trans_a, trans_b, &M, &N, &K, &alpha, &A[0, 0], &lda,
+               &A[0, py_start_voxel], &ldb, &beta, &C[0, py_start_epoch,0], &ldc)
+
+def compute_kernel_matrix(py_uplo, py_trans, py_n, py_k, py_alpha, py_a, int py_start_voxel, py_lda,
+                          py_beta, py_c, py_ldc):
+    """ use blas API wrapped by scipy.linalg.cython_blas to compute kernel matrix of SVM
+
+    The blas APIs process matrices in column-major, but our matrices are in row-major,
+    so we play the transpose trick here, i.e. A*B=(B^T*A^T)^T
+
+    In SVM with linear kernel, the distance of two samples
+    is essentially the dot product of them.
+    Therefore, the kernel matrix can be obtained by matrix multiplication.
+    Since the kernel matrix is symmetric, ssyrk is used,
+    the other half of the matrix is assigned later.
+    In our case, the dimension of samples is much larger than
+    the number samples, so we proportionally shrink the values of the kernel matrix
+    for getting more robust alpha values in SVM iteration.
+
+    Parameters
+    ----------
+    py_uplo: str
+    getting the upper or lower triangle of the matrix
+
+    py_trans: str
+    do transpose or not for the input matrix A
+
+    py_n: int
+    the row and column of the resulting matrix C
+    in our case, is num_epochs
+
+    py_k: int
+    the collapsed dimension of the multiplying matrices
+    i.e. the column of the first matrix after transpose if necessary
+    the row of the second matrix after transpose if necessary
+    in our case, is num_voxels
+
+    py_alpha: float
+    the weight applied to the input matrix A
+
+    py_a: 3D array in shape [num_assigned_voxels, num_epochs, num_voxels] in our case
+    the normalized correlation values of a voxel
+
+    py_start_voxel: int
+    the processed voxel
+    used to locate the input matrix A
+
+    py_lda: int
+    the stride of the input matrix A
+
+    py_beta: float
+    the weight applied to the resulting matrix C
+
+    py_c: 2D array in shape [num_epochs, num_epochs]
+    place to store the resulting kernel matrix
+
+    py_ldc: int
+    the stride of the resulting matrix
+
+    Returns
+    -------
+    py_c: 2D array in shape [num_epochs, num_epochs]
+    write the resulting kernel_matrix
+    for the processing voxel
+    """
+    cdef bytes by_uplo=py_uplo.encode()
+    cdef bytes by_trans=py_trans.encode()
+    cdef char* uplo = by_uplo
+    cdef char* trans = by_trans
+    cdef int N, K, lda, ldc
+    N = py_n
+    K = py_k
+    lda = py_lda
+    ldc = py_ldc
+    cdef float alpha, beta
+    alpha = py_alpha
+    beta = py_beta
+    cdef float[:, :, ::1] A
+    A = py_a
+    cdef float[:, ::1] C
+    C = py_c
+    blas.ssyrk(uplo, trans, &N, &K, &alpha, &A[py_start_voxel, 0, 0], &lda,
+               &beta, &C[0, 0], &ldc)
+    # shrink the values for getting more stable alpha values in SVM training iteration
+    py_c *= .001
```

Comment:
This is almost a fixed value. It gives us larger alpha values in SVM given our data shape (#features>>#samples). There is not much the users can tune about this.